Added cleanSavaData method - Suggested improvement

TeckniX opened the pull request:
The book makes a small mention of the DB schema:
`When defining fields for TranslateBehavior to translate, be sure to omit those fields from the translated model’s schema. If you leave the fields in, there can be issues when retrieving data with fallback locales.`

The issue is that if the model's schema has those fields, it will try to update them for whatever current language is selected. The above patch allows for only the 'default' language, as defined by `Config.language`, to be saved back into the model. If another language is selected at the time of the save, then the model's schema will not be updated with the data. Only the translate table data will be.

TeckniX committed:
Added cleanSavaData method - Suggested improvement

The book makes a small mention of the DB schema:
`When defining fields for TranslateBehavior to translate, be sure to omit those fields from the translated model’s schema. If you leave the fields in, there can be issues when retrieving data with fallback locales.`

The issue is that if the model's schema has those fields, it will try to update them for whatever current language is selected. The above patch allows for only the 'default' language, as defined by `Config.language`, to be saved back into the model. If another language is selected at the time of the save, then the model's schema will not be updated with the data. Only the translate table data will be.

dereuromark commented on lib/Cake/Model/Behavior/TranslateBehavior.php:
```diff
@@ -353,11 +354,35 @@ public function beforeSave(Model $Model, $options = array()) {
+			return true;
+		}
+		$fields = array_merge($this->settings[$Model->alias], $this->runtime[$Model->alias]['fields']);
+		if ($locale != Configure::read('Config.language')) {
```
!==

dereuromark commented on lib/Cake/Model/Behavior/TranslateBehavior.php:
```diff
@@ -353,11 +354,35 @@ public function beforeSave(Model $Model, $options = array()) {
+ * Prevents Model data from being saved if not in the 'base' language
+ * as defined by Config.language 
+ *
+ */
```
The doc block is missing the param and return declarations.

TeckniX committed:
Fixed docblock and check

As suggested by DerEuroMark, docblock added along with proper inequality

markstory commented on lib/Cake/Model/Behavior/TranslateBehavior.php:
```diff
@@ -353,11 +354,37 @@ public function beforeSave(Model $Model, $options = array()) {
+			foreach ($fields as $field) {
+				if (isset($Model->data[$Model->alias][$field])) {
+					unset($Model->data[$Model->alias][$field]);
+				}
```
You don't need this if. Unset() will not emit warnings.

TeckniX committed:
Removed unset()

Following Mark Story's comment about unset not triggering errors, it was removed

markstory commented:
Would we be able to get a test case for this? I'd like to wrap up as much of the 2.7 as possible this week so we can do an RC release.

TeckniX commented:
Sorry been busy and totally forgot to do so - I'll try to push it out today

Commit:
Added test case for default language save in model with translate field

Commit:
Added test case for default language save in model with translate field

TeckniX commented:
Still a n00b at unit test, let me know if it's not up to par.

markstory commented on lib/Cake/Test/Fixture/TranslateTitleFixture.php:
```diff
@@ -0,0 +1,58 @@
+ *
+ * @var string
+ */
+	public $table = 'i18n';
```
Won't this cause issues with the other translation table fixtures?

markstory commented:
Looks good to me. But I'm not a frequent use of the TranslateBehavior. Perhaps @ADmad, @dereuromark or @lorenzo can take a look?

ADmad commented:
@markstory I have actually never used the core TranslateBehavior :)  I use a shadow table based alternative.

lorenzo commented:
I will take a look at this in the evening

dereuromark commented:
Me neither.

lorenzo commented:
Ok, after reading the code it seems like this would bring the behavior a bit more inline to how it works in 3.0.

The problem is that it would break how it previously worked, which is something I'm not keen on doing.

TeckniX commented:
@lorenzo I don't believe it would break anything - It only checks for a matching field in the 'normal' table and prevents an over-write of the table data if it's not in the core language. 

Essentially, it allows you to add a translation table without having to touch your existing table, and it will keep it's core language updated. 

eg:
Posts table: id, title, content, created, modified
Post_i18ns: id, locale, model, foreign_key, field, content

If the default/core language is set to 'eng' (via locale) then on update the Posts table will be updated along with the Post_i18ns.
If the language set by locale is something else (like spa) then on update only the Post_i18ns table will be updated. 

As long as this is something that will be going into 3.0 then I'm all about it, but I don't believe the above would break the current 2.x release.

markstory commented:
@lorenzo ping - this is assigned to you.

dereuromark commented on lib/Cake/Model/Behavior/TranslateBehavior.php:
```diff
@@ -353,11 +354,35 @@ public function beforeSave(Model $Model, $options = array()) {
+	protected function _cleanSaveData(Model $Model) {
+		$locale = $this->_getLocale($Model);
+		if (empty($locale)) {
+			return true;
```
It should not return true then :)

TeckniX replied:
I must've missed the notification - Shouldn't it return true to assert that the data is clean?

dereuromark replied:
your doc block states return void, so why returning true?

TeckniX replied:
because I'm an idiot? :facepalm: 

I'll get it all fixed up. Thanks

markstory commented:
ping

markstory commented:
Closing as the 2.7. branch is coming to an end soon. If you'd like to pursue this further please reopen it against the `2.next` branch.
